### PR TITLE
ACT-FAST

### PR DESCRIPTION
**Pull request by guidoAI**
This is a new method to efficiently detect corners in large images, such as those obtained from the Parrot Bebop 1 and 2. The method is inspired by the method in the article: de Croon, G.C.H.E., and Nolfi, S. (2013, May). Act-corner: Active corner finding for optic flow determination. In Robotics and Automation (ICRA), 2013 IEEE International Conference on (pp. 4679-4684). IEEE.

The main idea of this particular implementation, called ACT-FAST, is that actively scanning the image allows to:
1. Skip uniform areas in the image. If these areas are contiguous, many non-corners will never be evaluated.
2. Follow edges. Typically, following an edge will bring an agent to a corner.

These two simple rules lead to a significant lower number of corner evaluations, while still detecting a reasonable number of corners. Moreover, since the agents scanning the image start on a grid, corners will be quite well-distributed over the image. Each step of the agent starts by classifying the agent location as a corner or not with FAST. For bigger images (e.g., 640 x 640), the computational advantage of ACT-FAST over the normal, exhaustive application of FAST becomes significant (in the order of a factor > 10). On a Bebop 2 I measured a processing time of ~0.003 seconds compared to ~0.04 seconds for exhaustive FAST corner checking. 

The code here is part of the following publication:
de Croon, G.C.H.E. "ACT-FAST: efficiently finding corners by actively exploring images.", in submission.

### CODE REVIEW

File: sw/airborne/modules/computer_vision/opticflow/opticflow_calculator.c [hunk 1]
```diff
@@ -307,23 +315,52 @@ bool calc_fast9_lukas_kanade(struct opticflow_t *opticflow, struct image_t *img,
...
+                   NULL);
+
+    }
+    else if (CORNER_METHOD == ACT_FAST) {
+        // ACT-FAST corner detection:
+        // TODO: all relevant things should be settings:
+        float long_step = 10; // 5
+        float short_step = 2; // 2
+        int min_gradient = 10; // 10
+        printf("opticflow->fast9_threshold = %d, n_agents = %d, n_time_steps = %d\n", opticflow->fast9_threshold, n_agents, n_time_steps);
```
Comment (kirkscheper) [inline sw/airborne/modules/computer_vision/opticflow/opticflow_calculator.c]: Remove or make optional

File: sw/airborne/modules/computer_vision/lib/vision/act_fast.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * The code here is part of the following publication:
+ * de Croon, G.C.H.E. "ACT-FAST: efficiently finding corners by actively exploring images.", in submission.
+ *
+ */
+
+#ifndef ACT_FAST_H
+#define ACT_FAST_H
+
+struct agent
+{
```
[the remark this answers is not in the document]
Comment (guidoAI) [reply sw/airborne/modules/computer_vision/lib/vision/act_fast.h]: ok - changed that

File: sw/airborne/modules/computer_vision/opticflow/opticflow_calculator.c (hunk 1 above)
Comment (guidoAI) [reply sw/airborne/modules/computer_vision/opticflow/opticflow_calculator.c]: made them settings and removed the printf

File: sw/airborne/modules/computer_vision/lib/vision/act_fast.c [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+  /*
+   * Procedure:
+   * 1) initialize agent positions
+   * 2) loop over the agents, moving and checking for corners
+   */
+
+  // method to determine the gradient:
+  // 0 = simple (-1, 0, 1)
+  // 1 = Sobel
+  int gradient_method = 1;
```
[the remark this answers is not in the document]
Comment (guidoAI) [reply sw/airborne/modules/computer_vision/lib/vision/act_fast.c]: made it a setting

**Comment by guidoAI** [comment]
I now added the relevant settings, and did a fix code style. It also runs on the bebop 2.

**Pull request merged**